flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/UserConfiguredOptionsFactory.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.contrib.streaming.state;
+
+import org.apache.flink.util.DynamicCodeLoadingException;
+import org.apache.flink.util.Preconditions;
+
+import org.rocksdb.ColumnFamilyOptions;
+import org.rocksdb.DBOptions;
+
+/**
+ * This is a encapsulation of user defined {@link OptionsFactory} through flink-conf.yaml.
+ * Keep in mind that before creating DBOptions and ColumnFamilyOptions,
+ * {@link #instantiateOptionsFactory(ClassLoader)} must be first called with given classloader.
+ */
+public final class UserConfiguredOptionsFactory implements OptionsFactory{
```

Comment:
I wonder why this class is required, couldn't we just do the `Class.forName(...)` directly in `RocksDBStateBackend::configure`? Is there any particular reason that we introduce this intermedate factory instead of just instanciating a factory immediately?

Reply:
`Class.forName(String name, boolean initialize, ClassLoader loader)` needs the `ClassLoader` which could only be accessed when creating the `RocksDBKeyedStateBackend`. I have to lazily    instantiate the options factory at that time. And I give this solution to wrap the class name first and then instantiate it when we could get the `ClassLoader`.